src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,454 @@
+
+#ifdef LIBRARY_MANAGER
+#include <sof/audio/module_adapter/module/intel_modules.h>
+#endif
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+
+/* 54cf5598-8b29-11ec-a8a3-0242ac120002 */
+
+DECLARE_SOF_UUID("lib_manager", lib_manager_uuid, 0x54cf5598, 0x8b29, 0x11ec,
+		 0xa8, 0xa3, 0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(lib_manager_tr, SOF_UUID(lib_manager_uuid), LOG_LEVEL_INFO);
+
+
+
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38*1024)
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(HOST_PAGE_SIZE)
+
+
+/**
+ * DMA buffer
+ */
+struct lib_manager_dma_buf {
+	uintptr_t w_ptr;	/**< write pointer */
+	uintptr_t r_ptr;	/**< read pointer */
+	uintptr_t addr;		/**< buffer start pointer */
+	uintptr_t end_addr;	/**< buffer end pointer */
+	uint32_t size;		/**< buffer size */
+	uint32_t avail;		/**< buffer avail data */
+};
+
+struct lib_manager_dma_ext {
+	struct dma_copy dma;
+	struct dma_chan_data *chan;
+	struct lib_manager_dma_buf dmabp;
+};
+
+struct ext_library loader_ext_lib;
+
+void *lib_manager_get_config(struct comp_dev *dev)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct module_data *md = &mod->priv;
+	struct module_config *dst = &md->cfg;
+
+	return dst->data;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+			mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
+	uint32_t inst_offset = instance_bss_size * HOST_PAGE_SIZE * instance_id;
+	uint8_t *va_base = (uint8_t *)(mod->segment[SOF_MAN_SEGMENT_BSS].v_base_addr + inst_offset);
+
+	tr_dbg(&lib_manager_tr,
+			"lib_manager_get_instance_bss_address() instance_bss_size: 0x%x, pointer: 0x%x",
+			instance_bss_size, va_base);
+	return va_base;
+}
+
+int lib_manager_allocate_module_instance(uint32_t module_id, uint32_t instance_id,
+						uint32_t is_pages, struct sof_man_module *mod)
+{
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			* HOST_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
+
+	if ((is_pages * HOST_PAGE_SIZE) > bss_size) {
+		tr_err(&lib_manager_tr,  "is_pages (%d) invalid, required: %d",
+				is_pages, bss_size/HOST_PAGE_SIZE);
+		return IPC4_FAILURE;
+	}
+	/*
+	 * TODO: Map bss memory here when mm_map() available.
+	 * And clear BSS
+	 */
+	va_base = va_base;
+	return 0;
+}
+
+uint32_t lib_manager_allocate_module(const struct comp_driver *drv,
+		   struct comp_ipc_config *ipc_config,
+		   void *ipc_specific_config)
+
+{
+	int ret = IPC4_FAILURE;
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
+	uint32_t entry_index = GET_MODULE_INDEX(module_id);
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
```

Comment:
can you please club all the struct definitions first followed by the rest for consistency with the rest of the code in the SOF repo